### ISSUE TITLE
Describe outputs of different Centerline extraction methods with more detail

### ISSUE THREAD

**Issue by NadiaBlostein** [spinalcordtoolbox/spinalcordtoolbox#4006]
# Feature justification
I think it would be easier for SCT developers to have a concise description of what objects are returned and files are written by the various Centerline extraction methods. For instance, the space and orientation in which objects & files are defined can be confusing and lead to downstream errors.

# Potential solution
Adding more detailed information about what objects are returned & files written in the comments section at the beginning each function. 

# Obtaining centerlines & understanding outputs

## Environment
* Python 3.8.15
* SCT: 5.8
* Shell: /bin/bash

## Data & Set up
_Note: unless otherwise specified, commands below are run from command-line._
	1. Download data [here](https://spinalcordtoolbox.com/user_section/tutorials/segmentation/before-starting.html)
	2. `cd single_subject/data/t2/`
	3. Segment spinal cord: `sct_deepseg_sc -i t2.nii.gz -c t2 -ofolder $PWD -qc $PWD/qc`

### A. `sct_get_centerline`
**Command:**
  ```
  sct_get_centerline -i t2.nii.gz -c t2
  ```
**Writes files:**
		* `t2_centerline.csv`: centerline coordinates defined in voxel ("pix") space and RPI orientation 
		* `t2_centerline.nii.gz`: centerline mask defined in voxel ("pix") space and input (e.g. _t2.nii.gz_) orientation (not RPI) 

### B. `spinalcordtoolbox.centerline.core.get_centerline`
**Command:**
  ```
  python
  from spinalcordtoolbox.centerline.core import get_centerline 
  from spinalcordtoolbox.centerline.core import ParamCenterline 
  from spinalcordtoolbox.image import * 
  from spinalcordtoolbox.types import Centerline
  param=ParamCenterline(algo_fitting = 'linear')
  im_centerline, arr_ctl, arr_ctl_der,fit_results = get_centerline(Image("t2_seg.nii.gz"),param=param)
  centerline = Centerline(arr_ctl[0],arr_ctl[1],arr_ctl[2],arr_ctl_der[0],arr_ctl_der[1],arr_ctl_der[2])
  ```
**Returns objects:**
	* `im_centerline`: centerline Image object defined in voxel ("pix") space and input orientation (e.g. _t2_seg.nii.gz_)
	* `arr_ctl`: centerline coordinates defined in voxel ("pix") space and RPI orientation
	* `centerline`: centerline object that stores points (centerline.points) defined in voxel ("pix") space and RPI orientation
	* `arr_ctl_der`
	* `fit_results`

### C. `spinalcordtoolbox.straightening._get_centerline`
**Command:**
```
python
from spinalcordtoolbox.straightening import _get_centerline 
from spinalcordtoolbox.centerline.core import get_centerline 
from spinalcordtoolbox.centerline.core import ParamCenterline 
from spinalcordtoolbox.types import Centerline
from spinalcordtoolbox.image import * 
param=ParamCenterline(algo_fitting = 'linear')
centerline = _get_centerline(Image("t2_seg.nii.gz"),param_centerline=param,verbose=1)
```
**Returns objects:**
	*`centerline`*: Centerline object that stores centerline derivatives (centerline.derivatives) and centerline coordinates (centerline.points) defined in physical ("phys") space and input (e.g. _t2_seg.nii.gz_) orientation

### D. `spinalcordtoolbox.types.save_centerline`
**Prerequisite:** generate Centerline object, either via `spinalcordtoolbox.straightening._get_centerline` or `spinalcordtoolbox.centerline.core.get_centerline`
**Command:**
```
python
centerline.save_centerline(fname_output='ctl')
centerline.save_centerline()
```
**Writes files:**
	*`ctl.npz`: File format that stores all of the Centerline object information (not just the coordinates)
	*`centerline.sct.npz`: default nomenclature if no `fname_output` given

**Comment by joshuacwnewton** [comment]
> **Note**: SCT is split into two main parts: the CLI and the API. To learn more about this, I would highly recommend reading the ["API code vs. CLI code"](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-API-code-vs-CLI-code) page on the SCT Developer Wiki first, since it helps to explain some key concepts that I'll be referring to in this comment.

I've been an SCT maintainer/developer for the past several years, so I can try to give some commentary that I hope will clarify things:

> ### A. `sct_get_centerline`

This is a CLI script, and should be the main way that most people generate a centerline image file. 

(Because it is a CLI script, it is easy to access the command via a terminal, and it's easy to add the command to research scripts. Additionally, you can still use this script to access some of the intermediate files by running this command with the `-r 0` option, too.)

> ### B. `spinalcordtoolbox.centerline.core.get_centerline`

This is an API function. It's the "internal" part of the `sct_get_centerline` script, but it's also used by other CLI scripts, too, when they require a centerline image (e.g. `sct_deepseg_sc`'s code calls `get_centerline`).

When you are writing code inside of the `spinalcordtoolbox/` package, and you require a centerline image for further processing, you should call `get_centerline` in your code, instead of calling the `sct_get_centerline` CLI script.

> ### C. `spinalcordtoolbox.straightening._get_centerline`

To be completely honest, this function is a bit weird, and its presence has confused me in the past. I don't think there should even be a `_get_centerline` function, and I think it would be good to refactor it into something else to make it less confusing. 

(Additionally, the function is prefixed with the `_` character, which in Python means "this function is private". In other words, it was designed to only be used in the context of the `straightening` module, and should not be used anywhere else.)

For these reasons, I would suggest to completely ignore this function for now. and only consider A and B.

> D. `spinalcordtoolbox.types.Centerline.save_centerline`

Admittedly, I have never worked with `Centerline` objects before, and I have never written code that calls the `save_centerline` method. (Even after 2+ years!)

Doing a search for the term `.save_centerline(` across the entire SCT codebase, it seems like this method is only used in 2 places. Because of how rare its usage is, IMO this method is not an important part of generating centerlines in SCT.

So, like before, I would suggest to ignore this method and use only A and B.

**Comment by NadiaBlostein** [comment]
Thanks a lot for the insights (I never thought of storing temp files by tweaking the `-r` option, thanks!)! To add on:

> ### C. `spinalcordtoolbox.straightening._get_centerline`

This is used by a few scripts (e.g. [spinalcordtoolbox.csa_pmj](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/spinalcordtoolbox/csa_pmj.py) and helped @rohanbanerjee and @benjamindeleener debug some code in the [template-generation code base](https://github.com/neuropoly/template/pull/13). 
One could indeed avoid `_get_centerline` and just stick to `get_centerline` (e.g. `_,arr_ctl,_,_ = spinalcordtoolbox.centerline.core.get_centerline`) by converting `arr_ctl` to physical ("phys") space (`spinacordtoolbox.image.transfo_pix2phys`). 

> ### D. `spinalcordtoolbox.types.Centerline.save_centerline`
So the main advantage of the `.save_centerline()` command is that it allows you to write a file that stores your Centerline object `self.derivatives`, `self.disks_levels`, `self.label_reference`) – whereas all the other approaches seem to only allow you to write a file with the Centerline coordinates (ie. `self.points`). Is this not the case?

## TLDR
@joshuacwnewton Do you agree that it would be worth specifying in the comments of the aforementioned functions what objects are output and files written (i.e. file extension, physical ["phys"] vs voxel ["pix"] space, orientation), or do you think that would just make things more confusing?

Thank you!

**Comment by joshuacwnewton** [comment]
> This is used by a few scripts (e.g. [spinalcordtoolbox.csa_pmj](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/spinalcordtoolbox/csa_pmj.py)) and helped @rohanbanerjee and @benjamindeleener debug some code in the [template-generation code base](https://github.com/neuropoly/template/pull/13).

Ah!! I didn't realize this function was being used elsewhere. Thank you for bringing this to my attention! My notifications are turned on for the spinalcordtoolbox repo, but not for side projects, so I'm often out of the loop for other efforts.

If the functionality in `straightening._get_centerline` is that useful for our research students, then I would suggest that we move the extra functionality to the main `get_centerline` function, and reduce the role that the `straightening` module plays in centerline creation. 

> So the main advantage of the `.save_centerline()` command is that it allows you to write a file that stores all of the attributes you have for your Centerline object (`self.points`, `self.derivatives`, `self.disks_levels`, `self.label_reference`) – whereas all the other approaches seem to only allow you to write a file with the Centerline coordinates (ie. `self.points`). Is this not the case?

You're absolutely right; the `Centerline` object does contain additional information that the other methods do not capture.

Truthfully, I've never worked with the extra fields that `Centerline` provides, so it's hard for me to say how important this additional information is. My impression was that saving this additional information wasn't necessary for the average user? But, I would love to be corrected here! 

> Do you agree that it would be worth specifying in the comments of the aforementioned functions what objects are output and files written (i.e. file extension, physical ["phys"] vs voxel ["pix"] space, orientation), or do you think that would just make things more confusing?

I think this kind of documentation would certainly be good to have in the docstrings of various methods/functions.

But, more importantly in my eyes, it would be good to have a unified interface for centerlines. (For example, if the extra fields and processing steps from C and D are important for research/development, I would want to see if we can fold this information into A and B, as opposed to maintaining 4 different ways of getting the centerline.)

**Comment by benjamindeleener** [comment]
I agree with the idea to move `_get_centerline`functionnalities to `get_centerline`. I usually work with physical coordinates when I use `Centerline` objects because it facilitates registration processes with multiple images/subjects, such as in the template generation.

The additional information that `Centerline` contains (`disks_levels`, `label_reference`, etc.) are very useful in some cases, and should be kept. However, it is true that saving `Centerline` objects with all the information is not something the average user will use.
